### ISSUE TITLE
Make mutable generic collection interfaces implement read-only collection interfaces

### ISSUE THREAD

**Issue by TylerBrinkley** [dotnet/runtime#31001]
## Rationale
It's long been a source of confusion that the mutable generic collection interfaces don't implement their respective read-only collection interfaces. This was of course due to the read-only collection interfaces being added after the fact and thus would cause breaking changes by changing a published interface API.

With the addition of default interface implementations in C#8/.NET Core 3.0 I think the mutable generic collection interfaces, `ICollection<T>`, `IList<T>`, `IDictionary<K, V>`, and `ISet<T>` should now implicitly inherit their respective read-only collection interfaces. This can now be done without causing breaking changes.

While it would have been nice for these interfaces to share members, I think the proposed API below is the best we can possibly do with the read-only interfaces being added after the fact.

As an added bonus, this should allow some simplification of the type checking in LINQ code to check for the read-only interfaces instead of the mutable interfaces.

## Proposed API

```diff
 namespace System.Collections.Generic {
-    public interface ICollection<T> : IEnumerable<T> {
+    public interface ICollection<T> : IReadOnlyCollection<T> {
-        int Count { get; }
+        new int Count { get; }
+        int IReadOnlyCollection<T>.Count => Count;
     }
-    public interface IList<T> : ICollection<T> {
+    public interface IList<T> : ICollection<T>, IReadOnlyList<T> {
-        T this[int index] { get; set; }
+        new T this[int index] { get; set; }
+        T IReadOnlyList<T>.this[int index] => this[index];
     }
-    public interface IDictionary<TKey, TValue> : ICollection<KeyValuePair<TKey, TValue>> {
+    public interface IDictionary<TKey, TValue> : ICollection<KeyValuePair<TKey, TValue>>, IReadOnlyDictionary<TKey, TValue> {
-        TValue this[TKey key] { get; set; }
+        new TValue this[TKey key] { get; set; }
-        ICollection<TKey> Keys { get; }
+        new ICollection<TKey> Keys { get; }
-        ICollection<TValue> Values { get; }
+        new ICollection<TValue> Values { get; }
-        bool ContainsKey(TKey key);
+        new bool ContainsKey(TKey key);
-        bool TryGetValue(TKey key, out TValue value);
+        new bool TryGetValue(TKey key, out TValue value);
+        TValue IReadOnlyDictionary<TKey, TValue>.this[TKey key] => this[key];
+        IEnumerable<TKey> IReadOnlyDictionary<TKey, TValue>.Keys => Keys;
+        IEnumerable<TValue> IReadOnlyDictionary<TKey, TValue>.Values => Values;
+        bool IReadOnlyDictionary<TKey, TValue>.ContainsKey(TKey key) => ContainsKey(key);
+        bool IReadOnlyDictionary<TKey, TValue>.TryGetValue(TKey key, out TValue value) => TryGetValue(key, out value);
     }
-    public interface ISet<T> : ICollection<T> {
+    public interface ISet<T> : ICollection<T>, IReadOnlySet<T> {
-        bool IsProperSubsetOf(IEnumerable<T> other);
+        new bool IsProperSubsetOf(IEnumerable<T> other);
-        bool IsProperSupersetOf(IEnumerable<T> other);
+        new bool IsProperSupersetOf(IEnumerable<T> other);
-        bool IsSubsetOf(IEnumerable<T> other);
+        new bool IsSubsetOf(IEnumerable<T> other);
-        bool IsSupersetOf(IEnumerable<T> other);
+        new bool IsSupersetOf(IEnumerable<T> other);
-        bool Overlaps(IEnumerable<T> other);
+        new bool Overlaps(IEnumerable<T> other);
-        bool SetEquals(IEnumerable<T> other);
+        new bool SetEquals(IEnumerable<T> other);
// Adding this new member is required so that there's a most specific Contains method on ISet<T> since ICollection<T> and IReadOnlySet<T> define it too
+        new bool Contains(T value) => ((ICollection<T>)this).Contains(value); 
+        bool IReadOnlySet<T>.Contains(T value) => ((ICollection<T>)this).Contains(value);
+        bool IReadOnlySet<T>.IsProperSubsetOf(IEnumerable<T> other) => IsProperSubsetOf(other);
+        bool IReadOnlySet<T>.IsProperSupersetOf(IEnumerable<T> other) => IsProperSupersetOf(other);
+        bool IReadOnlySet<T>.IsSubsetOf(IEnumerable<T> other) => IsSubsetOf(other);
+        bool IReadOnlySet<T>.IsSupersetOf(IEnumerable<T> other) => IsSupersetOf(other);
+        bool IReadOnlySet<T>.Overlaps(IEnumerable<T> other) => Overlaps(other);
+        bool IReadOnlySet<T>.SetEquals(IEnumerable<T> other) => SetEquals(other);
+    }
 }
```

## Binary Compatibility Test
I was able to test that this change doesn't break existing implementers with the following custom interfaces and by simply dropping the new interfaces dll to the publish folder without recompiling the consuming code, the `IMyReadOnlyList<T>` interface was automatically supported without breaking the code.

### Original Interfaces DLL code
```c#
namespace InterfaceTest
{
    public interface IMyReadOnlyList<T>
    {
        int Count { get; }
        T this[int index] { get; }
    }

    public interface IMyList<T>
    {
        int Count { get; }
        T this[int index] { get; set; }
    }
}
```

### New Interfaces DLL code
```c#
namespace InterfaceTest
{
    public interface IMyReadOnlyList<T>
    {
        int Count { get; }
        T this[int index] { get; }
    }

    public interface IMyList<T> : IMyReadOnlyList<T>
    {
        new int Count { get; }
        new T this[int index] { get; set; }
        int IMyReadOnlyList<T>.Count => Count;
        T IMyReadOnlyList<T>.this[int index] => this[index];
    }
}
```

### Consuming Code
```c#
using System;
using System.Collections.Generic;

namespace InterfaceTest
{
    class Program
    {
        static void Main()
        {
            var myList = new MyList<int>();
            Console.WriteLine($"MyList<int>.Count: {myList.Count}");
            Console.WriteLine($"IMyList<int>.Count: {((IMyList<int>)myList).Count}");
            Console.WriteLine($"IMyReadOnlyList<int>.Count: {(myList as IMyReadOnlyList<int>)?.Count}");
            Console.WriteLine($"MyList<int>[1]: {myList[1]}");
            Console.WriteLine($"IMyList<int>[1]: {((IMyList<int>)myList)[1]}");
            Console.WriteLine($"IMyReadOnlyList<int>[1]: {(myList as IMyReadOnlyList<int>)?[1]}");
        }
    }

    public class MyList<T> : IMyList<T>
    {
        private readonly List<T> _list = new List<T> { default, default };

        public T this[int index] { get => _list[index]; set => _list[index] = value; }

        public int Count => _list.Count;
    }
}
```

### Original Output
```
MyList<int>.Count: 2
IMyList<int>.Count: 2
IMyReadOnlyList<int>.Count:
MyList<int>[1]: 0
IMyList<int>[1]: 0
IMyReadOnlyList<int>[1]:
```

### New Output
```
MyList<int>.Count: 2
IMyList<int>.Count: 2
IMyReadOnlyList<int>.Count: 2
MyList<int>[1]: 0
IMyList<int>[1]: 0
IMyReadOnlyList<int>[1]: 0
```

Moved from dotnet/runtime#16151

## Updates
* Added `ISet<T>` implementing `IReadOnlySet<T>` since https://github.com/dotnet/runtime/issues/2293 has been implemented.

**Comment by TylerBrinkley** [comment]
@safern @GrabYourPitchforks @terrajobst Is this something we could get triaged soon? I'd love for this to make it into .NET 5.

**Comment by safern** [comment]
@eiriktsarpalis and @layomia are the new System.Collections owners so I'll defer to them.

**Comment by TylerBrinkley** [comment]
> > @terrajobst
> > > We can't make `ISet<T>` extend `IReadOnlySet<T>` (for the same reason that we couldn't do for the other mutable interfaces).
> > 
> >
> > Is this still true even with default interface methods? Does that mean [dotnet/corefx#41409](https://github.com/dotnet/corefx/issues/41409) should be closed?
>
> We discussed this. We used to think that that DIMs _would_ work, but when we walked the solution we concluded that it would result commonly in a shard diamond which would result in an ambiguous match. However, this was recently challenged so I think I have to write it down and make sure it's actually working or not working.

Here's a comment @terrajobst had about this in a separate issue, https://github.com/dotnet/runtime/issues/2293#issuecomment-579524652

Hoping for some more insight.

**Comment by TylerBrinkley** [comment]
I updated the proposal to add a DIM `Contains` method to `ISet<T>` as the current `Contains` method is defined on `ICollection<T>` and so if you tried to call `Contains` on an `ISet<T>` the call would be ambiguous with the `ICollection<T>` and `IReadOnlySet<T>` versions. The new method would take precedence in this case and it's implementation if not overridden would delegate to `ICollection<T>`'s implementation as `IReadOnlySet<T>`'s does. It's not ideal but this is the only incidence of such a solution where this is required for the proposal.

**Comment by TylerBrinkley** [comment]
As an alternative since `IReadOnlySet<T>` hasn't released yet, instead of adding another `Contains` method to the list including `ICollection<T>`, `IReadOnlySet<T>`, and now `ISet<T>` we could instead add an invariant read-only collection interface as proposed in https://github.com/dotnet/runtime/issues/30661 and have it implement a `Contains` method and then have `IReadOnlySet<T>` implement this new invariant interface and remove it's `Contains` method.

With that change then `ICollection<T>` could implement this new invariant read-only collection interface.

So instead of this.

```c#
namespace System.Collections.Generic {
    public interface ICollection<T> : IReadOnlyCollection<T> {
        bool Contains(T value);
        ...
    }
    public interface IReadOnlySet<T> : IReadOnlyCollection<T> {
        bool Contains(T value);
        ...
    }
    public interface ISet<T> : IReadOnlySet<T> {
        new bool Contains(T value) => ((ICollection<T>)this).Contains(value);
        bool ICollection<T>.Contains(T value);
        bool IReadOnlySet<T>.Contains(T value) => ((ICollection<T>)this).Contains(value);
        ...
    }
}
```

it would be this.

```c#
namespace System.Collections.Generic {
    public interface IReadOnlyCollectionInvariant<T> : IReadOnlyCollection<T> {
        bool Contains(T value);
        ...
    }
    public interface ICollection<T> : IReadOnlyCollectionInvariant<T> {
        new bool Contains(T value);
        bool IReadOnlyCollectionInvariant<T>.Contains(T value) => Contains(value);
        ...
    }
    public interface IReadOnlySet<T> : IReadOnlyCollectionInvariant<T> {
        ...
    }
}
```

**Comment by mikernet** [comment]
As a follow-up to this:

Perhaps LINQ methods that have optimized execution paths on interfaces like `IList<T>` could be updated to optimize on `IReadOnlyList<T>` instead since all `IList<T>` will implement `IReadOnlyList<T>` and then new code can begin moving to a sane development model where custom read-only collections don't need to extraneously implement the writable interfaces as well. I recall the justifications for optimizing LINQ on writable interfaces instead of read-only interfaces were:

1. `IReadOnlyList<T>` is covariant and thus slower.
2. `IReadOnlyList<T>` hasn't been around as long as `List<T>` so some code may not implement it and thus not get the optimizations.
3. It is "standard" to implement both `IList<T>` and `IReadOnlyList<T>` on all list-like collections anyway, I believe mostly because of the perf/optimization issues above?

I could be wrong but I believe there have been runtime updates to address point 1. Point 2 would be addressed by this issue. Point 3 is something that we could move away from with these changes to make read-only collection interfaces real first-class citizens in the framework that are actually usable.

A remaining sore point would be the annoying lack of support for read-only interfaces in many places where they should be supported. For example, `new List<T>(IReadOnlyList<T> source)` does not exist and adding it is problematic because of the diamond problem - many collections currently implement both `IList<T>` and `IReadOnlyList<T>` directly so overload resolution would fail. This could be addressed in the BCL by updating writable collections to only directly implement the writable interface but might be confusing if people use other collections that haven't been updated as such.

Collections are by far one of the most poorly designed areas of the BCL and fundamental to development so cleaning this up as much as possible with a bit of help from DIMs would be really nice.

**Comment by vpenades** [comment]
I would like to propose an alternative solution:

The introduction of a new interface for each type, that includes the two original interfaces, for example:

```c#
IList<T>
IReadOnlyList<T>
List<T> : IList<T> , IReadOnlyList<T>
```

What I propose is this:

```c#
IList<T>
IReadOnlyList<T>
IWriteableList<T> : IList<T>, IReadOnlyList<T>
List<T> : IWriteableList<T>
```

So, from the point of view of the runtime, it would just require introducing a new interface, so I don't think it would break any compatibility. And developers would progressively move from using IList<T> to IWriteableList<T>

**Comment by jhudsoncedaron** [comment]
So it turns out this is a very peanut-buttery performance optimization as well. There are a number of places in Linq that check for IEnumerable<T> for ICollection<T> or IList<T> for performance improvements and the report is that these can't be changed to check both for performance reasons; yet I found place after place in my own code where only IReadOnlyCollection/List<T> is provided and a mutable collection would be meaningless. I also found many places where ICollection/List<T> is dynamically downcast to its ReadOnly variant with an actual synthesis if the downcast fails.

Particularly to the implemenation of .Select, I found that making this change to that all ICollection/List<T> also provide their IReadOnly variants yields 33-50% performance improvement at only the cost of JIT time (to provide the interfaces where they are not). I also found that adding an IReadOnlyCollection.CopyTo() is a possible improvement but the gains aren't needed because implementing IListSource on the projection returned by Select() on an IReadOnlyCollection yields more gain than that ever could.

Synthetic benchmark attached showing the component changes. The first number is garbage (ensures the memory is paged in); the second and third numbers show the baseline, and the last number shows the gains. (It's approximately equal to the second number, which means I found all the performance gains). I found on my machine about every third run is garbage due to disturbance, so run it three times and keep the closest two numbers.
[selectbenchmarks.zip](https://github.com/dotnet/runtime/files/6971050/selectbenchmarks.zip)

**Comment by jhudsoncedaron** [comment]
Update: Saying that IReadOnlyCollection.CopyTo() was not needed because of a better implementation involving IListSource() proved to be overfitting to the test.

It turns out that there is more cheap (but not free) performance improvements lying around for direct invocation of .ToList() and .ToArray() (commonly used for shallow copies of arrays), List.AddRange(), List.InsertRange(), and List's own Constructor to the point where adding `IReadOnlyCollection<T>.CopyTo(T array, int offset);` is worthwhile. There is no gain for the default implementation but a custom implementation providing `.CopyTo()` yields cheap performance gains.

The default implementation would be:

````
public partial interface void IReadOnlyCollection<T>.CopyTo(T[] target, int offset) {
    if (this is ICollection<T> coll)  { coll.CopyTo(target, offset); return ; }
    foreach (var item in this) target[offset++] = item;
    return;
}
````

**Comment by jhudsoncedaron** [comment]
@eiriktsarpalis : I found that to be not an alternative design but a different design defect that could be fixed independently. `ICollection<T> : IReadOnlyCollection<T>` is not the heart of this one but rather `IList<T> : IReadOnlyList<T>`, `IDictionary<T> : IReadOnlyDictionary<T>`, and `ISet<T> => IReadOnlySet<T>`.

**Comment by jhudsoncedaron** [comment]
Not implementing this is slowly causing us more and more problems. The root is `IList<T>` and `IReadOnlyList<T>` are conflicting overloads (along with `ICollection<T>` and `IReadOnlyCollection<T>`). We have a large number of implementations of these interfaces and a decent number of extension methods on lists. The conflicting overloads are causing chronic problems.

It almost feels like a dumb compiler, but it's not. It doesn't matter which way half of the the overloads are resolved. The implementations are copy/paste of each other.

**Comment by mikernet** [comment]
@jhudsoncedaron This could potentially alleviate some of those issues:

https://github.com/dotnet/csharplang/discussions/4867

Would still be nice to fix collections though.

**Comment by jhudsoncedaron** [comment]
@mikernet : It won't. To work I need it to do exactly what it says it won't do: "Overload priorities should only be considered when disambiguating methods within the same assembly". I need to win (or in one case lose) against overloads shipped with .NET itself.

**Comment by mikernet** [comment]
Not for the example you showed you don't. Your ICollection or IReadOnlyCollections WILL beat the IEnumerable overload from the BCL.

**Comment by jhudsoncedaron** [comment]
@mikernet : Breaking example: `IReadOnlyDictionary<TK, TV>.GetValueOrDefault(TK key, TV default)`. Discovered in our codebase where these two had been hanging out for some time until somebody tried to call one of them directly on a `Dictionary<TK, TV>`.

**Comment by mikernet** [comment]
I didn't say it will solve all your problems. I said it could alleviate some of the ones you showed.

Also, regarding:

> One of our programmers really likes calling `.Any()` for readability as opposed to `.Count > 0`

There are several analyzers that detect this. I don't know off hand which rule we use - probably style cop or roslynator. Just set a company-wide analyzer rule to warn when someone does that and then he can't merge code that violates the analyzer rule - problem solved :P It's fairly well known that LINQ methods shouldn't be preferred over faster alternatives when available.

**Comment by Bip901** [comment]
This is a highly requested feature not only for performance reasons but also for object oriented-ness and greatly simplified code. It has been requested in tens of issues in this repository, some of them mentioned here.

Will we be seeing this feature in .NET 7?

**Comment by joshudson** [comment]
@Bip901: I could make a PR immediately; but the API review team is unwilling to approve the API change.

**Comment by eiriktsarpalis** [comment]
It's not an issue of unwillingness. As it stands, it seems unlikely we would consider augmenting existing interfaces with DIMs, because this is known to create both compile-time and runtime breaking changes in downstream consumers due to the potential for diamond ambiguities. See https://github.com/dotnet/runtime/issues/52775#issuecomment-843446017 for an example.

Related to #55106.

**Comment by Joe4evr** [comment]
@eiriktsarpalis As an alternative, is it worth shipping an analyzer that detects types implementing `ICollection<T>`/`IList<T>` without the `ReadOnly` versions, and suggests to add them to the type declaration? (If I'm not mistaken, the members in read-only interfaces are a strict sub-set of the writable versions, so it wouldn't require adding new members that the type didn't already have.)

**Comment by Timovzl** [comment]
> As an alternative, is it worth shipping an analyzer that detects types implementing ICollection<T>/IList<T> without the ReadOnly versions, and suggests to add them to the type declaration?

I love this idea as an alternative. While it does not make for a foolproof solution, it should decrease the chance of encountering missing readable interfaces over time.

**It would be especially great if .NET's own classes started adhering to the proposed analyzer.** For example, consider the concrete [Grouping](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Grouping.cs#L50) type used by LINQ's `ToLookup()`. It implements `IList<T>` but not `IReadOnlyList<T>`. For one thing, this hinders optimizations by library authors. As an example, one might optimize for `IReadOnlyList<T>` (can index instead of enumerating) or `IReadOnlyCollection<T>` (can use multiple enumeration if necessary). However, separately optimizing for readable _and_ writable interfaces tends to introduce an overgrowth of special cases and branches.

**Comment by Bip901** [comment]
That's a good alternative. I'd still prefer a real solution. These interfaces are so widespread that they deserve a breaking change (then again, you can use the same fact to argue against breaking them).

**Comment by terrajobst** [comment]
Analyzers aren't a viable alternative for breaking changes in highly used APIs because it would likely immediately break the entire library ecosystem.

**Comment by jhudsoncedaron** [comment]
"it would likely immediately break the entire library ecosystem."

With this specific change, I'm anticipating few if any breaking changes in practice from the added default methods. While trying to add new methods to `IEnumerable` and overriding them below does not work, the tight-in checks that mimic this sequence of changes don't fail in the same way.

The next-best alternative is adding `IList2 : IList, IReadOnlyList` and changing hundreds of properties over to using it thus taking hundreds of breaking changes. The weight of this design defect does not stop growing.

**Comment by terrajobst** [comment]
The problem, and it has been a while that I looked, was that there were diamond dependencies in how customers implement collection interfaces which would result in `TypeLoadException`s because the runtime can't decide which default implementation to use. At compile time, it would fail to compile, forcing the developer to pick, but at runtime it's basically a fatal crash.

**Comment by TylerBrinkley** [comment]
> It's not an issue of unwillingness. As it stands, it seems unlikely we would consider augmenting existing interfaces with DIMs, because this is known to create both compile-time and runtime breaking changes in downstream consumers due to the potential for diamond ambiguities. See [#52775 (comment)](https://github.com/dotnet/runtime/issues/52775#issuecomment-843446017) for an example.
> 
> Related to #55106.

@eiriktsarpalis I do not think your diamond problem example is analagous to what's proposed here. In your example you added a completely new member and defined multiple DIM's for it across different interfaces that do not inherit from each other. That is quite different to this proposal which does not add any new members and only has a single DIM defined for each member.

Everytime I look at this I fail to see how a diamond could occur. Please if anyone has a true example of how a diamond would occur with this proposal I'd love to know.

**Comment by jhudsoncedaron** [comment]
I too have utterly failed to reproduce the binary load problem. Here is a much more complete demo of the changes proposed and I could not construct a reasonable failure case.

Attached is a sample set of projects. To test, build CollectionDemo, Beta\CollectionInterfaces and Gamma\CollectionInterfaces. Run the demo binary, then copy the dll from Beta or Gamma over the dll in the name name in CollectionDemo and run. CollectionDemo contains some tests for List and Dictionary to force them to be loaded and at the very end a precise check to see if the change was applied correctly.

[CollectionInterfaces.zip](https://github.com/dotnet/runtime/files/8227011/CollectionInterfaces.zip)

Changeset BETA is a mirror of the cautious proposal here. It applied without load errors as expected.
Changeset GAMMA is an aggressive change hunting out performance problems in Linq and List.InsertRange. To my surprise, it also applied without load errors.

Summary code:
````
    public interface IReadOnlyCollection<out T> : IEnumerable<T>
    {
        int Count { get; }
    }

#if GAMMA
    public interface IReadOnlyCollectionExact<T> : IReadOnlyCollection<T>
    {
        bool Contains(T item) { /* snip */ }
        void CopyTo(T[] array, int arrayIndex) { /* snip */ }
    }
#endif

    public interface IReadOnlyList<out T> : IReadOnlyCollection<T>
    {
        T this[int index] { get; }
    }

    public interface IReadOnlyDictionary<TK, TV> : IReadOnlyCollection<System.Collections.Generic.KeyValuePair<TK, TV>>
#if GAMMA
        , IReadOnlyCollectionExact<System.Collections.Generic.KeyValuePair<TK, TV>>
#endif
    {
        bool TryGetValue(TK key, out TV value);
        TV this[TK key] { get; }
        IEnumerable<TK> Keys { get; }
        IEnumerable<TV> Values { get; }

#if GAMMA
        void IReadOnlyCollectionExact<System.Collections.Generic.KeyValuePair<TK, TV>>.CopyTo(System.Collections.Generic.KeyValuePair<TK, TV>[] array, int arrayIndex)
            => ((ICollection<System.Collections.Generic.KeyValuePair<TK, TV>>)this).CopyTo(array, arrayIndex);

        bool IReadOnlyCollectionExact<System.Collections.Generic.KeyValuePair<TK, TV>>.Contains(System.Collections.Generic.KeyValuePair<TK, TV> item)
            => this.TryGetValue(item.Key, out var value) && object.Equals(item.Value, value);
#endif
    }

    public interface ICollection : IEnumerable
    {
        int Count { get; }
        bool IsSynchronzied { get; }
        object SyncRoot { get; }
    }

    public interface ICollection<T> : IEnumerable<T>
#if BETA
        , IReadOnlyCollection<T>
#endif
#if GAMMA
        , IReadOnlyCollectionExact<T>
#endif
    {
        int Count { get; }
        void Add(T item);
        void Clear();
        bool Contains(T item);
        void CopyTo(T[] array, int arrayIndex);
        bool Remove(T item);

#if BETA
        int IReadOnlyCollection<T>.Count => Count;
#endif
#if GAMMA
        bool IReadOnlyCollectionExact<T>.Contains(T item) => Contains(item);
        void IReadOnlyCollectionExact<T>.CopyTo(T[] array, int arrayIndex) => CopyTo(array, arrayIndex);
#endif
    }

    public interface IList<T> : ICollection<T>
#if BETA
        , IReadOnlyList<T>
#endif
    {
        T this[int index] { get; set; }
        void Insert(int index, T item);
        void RemoveAt(int index);

#if BETA
        int IReadOnlyCollection<T>.Count => Count;
        T IReadOnlyList<T>.this[int index] => this[index];
#endif
    }

    public interface IDictionary<TK, TV> : ICollection<System.Collections.Generic.KeyValuePair<TK, TV>>
#if BETA
        , IReadOnlyDictionary<TK, TV>
#endif
    {
        TV this[TK key] { get; set; }
        ICollection<TK> Keys { get; }
        ICollection<TV> Values { get; }
        void Add(TK key, TV value);
        bool ContainsKey(TK key);
        bool TryGetValue(TK key, out TV value);
        bool Remove(TK key);

#if BETA
        int IReadOnlyCollection<System.Collections.Generic.KeyValuePair<TK, TV>>.Count => Count;
        TV IReadOnlyDictionary<TK, TV>.this[TK key] => this[key];
        IEnumerable<TK> IReadOnlyDictionary<TK, TV>.Keys => Keys;
        IEnumerable<TV> IReadOnlyDictionary<TK, TV>.Values => Values;
        bool IReadOnlyDictionary<TK, TV>.TryGetValue(TK key, out TV value) => TryGetValue(key, out value);
#endif
    }
````

If there is a failure mode it should be possible to mutate CollectionDemo into triggering it while still keeping CollectionInterfaces matching this proposal. (I'm operating on a cut-down version to save time. I should have all cases covered, but I might not.)

I could push this farther. There's a guy who wants a non-generic `IReadOnlyCollection` inserted directly beneath `IEnumerable`. (Alternate proposal to `TryGetNonEnumeratedCount`.) After seeing changeset GAMMA work I'm convinced even that could be made to work.

**Comment by MichalStrehovsky** [comment]
Consider following valid user code:

```csharp
interface IMyDefaultReadOnlyCollection<T> : IReadOnlyCollection<T>
{
    int IReadOnlyCollection<T>.Count => 0;
}

class MyCollection<T> : ICollection<T>, IMyDefaultReadOnlyCollection<T>
{
    // Adding implementation of ICollection<T> left as an exercise
    // IReadOnly is left default-implemented from IMyDefaultReadOnlyCollection
}
```

Now suppose we add the `IReadOnlyCollection<T>` implementation to `ICollection<T>`. The implementation of `IReadOnlyCollection<T>.Count` on `MyCollection<T>` is now ambiguous between the implementation provided by `IMyDefaultReadOnlyCollection<T>` (old) and `ICollection<T>` (added in the breaking change). The class won't compile or work at runtime (if compiled against old BCL).

There are _very few_ scenarios where default interface methods are non-breaking. The rules are basically captured in https://github.com/dotnet/corefx/pull/41949.

**Comment by Bip901** [comment]
It's true that this is a breaking change, but the benefits far outweigh the relatively rare breaking cases. Better fix this API late than never.

**Comment by vpenades** [comment]
> It's true that this is a breaking change, but the benefits far outweigh the relatively rare breaking cases. Better fix this API late than never.

What about nuget packages that are no longer maintained but are critical for a number of projects?  unless the "relatively rare cases" can be quantified, I think this change might be beneficial for some developers, and extremely harmful for others.

Also, this change would be added to the newer frameworks, most probably Net7, so it would introduce a behavior conflict between net7 and older frameworks, specially netstandard2.0 which is widely used as the baseline by the majority of nuget packages.

**Comment by Bip901** [comment]
> What about nuget packages that are no longer maintained but are critical for a number of projects? 

Then those projects shouldn't migrate to .NET 7 until they've migrated their dependencies, just like projects couldn't migrate to .NET Core while they had .NET Framework dependencies.
Microsoft is known as a company that excels in backwards compatibility, but sometimes it's beneficial to make bold decisions.

> unless the "relatively rare cases" can be quantified

Perhaps we could search a repository of libraries for explicit implementations of readonly interfaces (e.g. regex `IReadOnlyCollection<.+>\.Count`).

**Comment by Bip901** [comment]
Problematic wording aside, @jhudsoncedaron is correct: the regex matched 109 files, but they also need to be an interface/base class of another class and that other class needs to implement ICollection<T> as well. Overall, **very** rare breaking cases.

**Comment by astrowalker** [comment]
It would be good to take historic view on this as well. There is some present codebase for sure and breaking change will have additional cost for it. But will new codebase be added or not? I guess it will, so every day of postponing fixing the issue (after all it should be done from the very beginning) is making all incoming codebase developers pay the price. It is another "make" history "I already have 3 users" only with turning point shifted forward.

**Comment by fitdev** [comment]
I hope it gets implemented ASAP, especially now that there have been DIMs for a few years.

**Comment by Daniel-Svensson** [comment]
@terrajobst  @stephentoub 

Would you consider making this change given that the type loader problem is solved ?  

If so do you have any thoughts on what would be an acceptable solution (for example using som kind of internal mechanism such as a special attribute to give internal overloads lower priority)

**Comment by zms9110750** [comment]
If  don't want to modify the code, how about adding a new namespace directly?
```csharp
internal interface IReadOnlyCollection<out T> : IEnumerable<T>, IEnumerable
{
	int Count { get; }
	bool IsReadOnly { get; }
}

internal interface ICollection<T> : IReadOnlyCollection<T>
{
	void Add(T item);

	void Clear();

	bool Contains(T item);

	void CopyTo(T[] array, int arrayIndex);

	bool Remove(T item);
}
```
Compared to having a bool member in the Readonly collection, I am even more surprised why I am not allowed to use Add in an interface with an Add method.

In our language philosophy, interfaces only specify what the minimum can do. I haven't seen any definition and decided what an interface is not allowed to do.In terms of responsibility extension, read write is a perfect extension for read-only.

And, in our actual use, why do we give someone a read-only collection? Because I want him to see my changes and passively accept them when I make changes, but he can observe the changes.But today, either we use a specific type. Either create a new instance.
The former does not conform to abstract programming, while the latter cannot meet the requirements.Compared to these two, I think the 'only' of readonly can be ignored.

**Comment by jhudsoncedaron** [comment]
@zms9110750 : The problem with this idea is the shear number of breaking changes required to switch almost everything over to the new interfaces. Same reason I didn't just make a parallel interface set.

**Comment by huoyaoyuan** [comment]
We should also include the non-generic interfaces (#91043) in discussion. Basically they are suffering the same risk of breaking.

**Comment by Daniel-Svensson** [comment]
Great to se this being reviewed for NET9.

I have a working example of solving the loader peoblem (dependency diamond priblem) so it is possible to add DIM to already defined interface methods without making it a breaking change at https://github.com/dotnet/runtime/compare/main...Daniel-Svensson:runtime:default_interface_method_fallbacks 

The POC reuse an existing attribute for simplicity, it should be replaced with something new instead. 

The POC makes t[he following solution]( [run](https://github.com/Daniel-Svensson/readonlycollection) based on 
@MichalStrehovsky example run as expected (with current runtime it throws).

@terrajobst do you want to handle the runtime problem as part of this issue or as a separate issue or pr?

There are several  possible such as a a new simplw atrribute (internal to runtime) , attribute with priority and/or maybe inferring some priority from where the DIM is defined (in assembly as interfacemethod overridden, in separate assembly "3rs party" , in same assembly as the concrete class). The POC is based on the first simple possible fix I could think of, where an attribute marks a method as "low priority"/fallback to use id no other implementation is found).

To keep being able to use the solution in future NET versions it might make sense to make it a non public type and maybe even enforce that it is only used in the same assembly as the interface method being implemented (to prevent other libraries for using it on runtime types).

**Comment by Daniel-Svensson** [comment]
@TylerBrinkley maybe you want to just add a short section to the issue mentioning that the diamond DIM problem need to be solved. You don't have to mention my POC or possible solutions to that problem, but it *would be good to mention that it should be solvable and not block the suggestion*.

**Comment by terrajobst** [comment]
I have created a repo trying to replicate the shared diamond problem and it doesn't seem to exist:

https://github.com/terrajobst/ReadOnlyCollectionInterfaces

I suggest we put this work on early for .NET 9.0 and see whether anything shows up. So far, it looks promising.

**Comment by MichalStrehovsky** [comment]
> I have created a repo trying to replicate the shared diamond problem and it doesn't seem to exist:
> 
> https://github.com/terrajobst/ReadOnlyCollectionInterfaces
> 
> I suggest we put this work on early for .NET 9.0 and see whether anything shows up. So far, it looks promising.

I've submitted two PRs to the repo that show two different failure modes - can be useful for discussions around the extent of this breaking change. If someone has other ideas for how this could break, it would probably be good to collect everything in the form of PRs to the repo.

* https://github.com/terrajobst/ReadOnlyCollectionInterfaces/pull/1
* https://github.com/terrajobst/ReadOnlyCollectionInterfaces/pull/2

This could be used to e.g. scope out how many things would be broken by scanning NuGet for these patterns. Ultimately we'd learn the most by trying this very early in .NET 9 so that we have runway to roll back if needed.

This would be better to land at or before https://github.com/dotnet/csharplang/blob/main/proposals/ref-struct-interfaces.md language feature lands because at that point adding instance default interface methods to existing interfaces becomes universally source+binary breaking (search for "default interface methods" in that doc).

**Comment by jhudsoncedaron** [comment]
Higher up in the thread somebody posted how to patch the loader to get it to auto-resolve those diamonds with a hinting attribute. So you can in fact reduce the breakage points to zero.

**Comment by bartonjs** [comment]
It sounds like we're ready to try the experiment.  To everyone who is celebrating right now, remember: if it goes poorly we're going to back it out :smile:.

```diff
 namespace System.Collections.Generic {
-    public interface ICollection<T> : IEnumerable<T> {
+    public interface ICollection<T> : IReadOnlyCollection<T> {
-        int Count { get; }
+        new int Count { get; }
+        int IReadOnlyCollection<T>.Count => Count;
     }
-    public interface IList<T> : ICollection<T> {
+    public interface IList<T> : ICollection<T>, IReadOnlyList<T> {
-        T this[int index] { get; set; }
+        new T this[int index] { get; set; }
+        T IReadOnlyList<T>.this[int index] => this[index];
     }
-    public interface IDictionary<TKey, TValue> : ICollection<KeyValuePair<TKey, TValue>> {
+    public interface IDictionary<TKey, TValue> : ICollection<KeyValuePair<TKey, TValue>>, IReadOnlyDictionary<TKey, TValue> {
-        TValue this[TKey key] { get; set; }
+        new TValue this[TKey key] { get; set; }
-        ICollection<TKey> Keys { get; }
+        new ICollection<TKey> Keys { get; }
-        ICollection<TValue> Values { get; }
+        new ICollection<TValue> Values { get; }
-        bool ContainsKey(TKey key);
+        new bool ContainsKey(TKey key);
-        bool TryGetValue(TKey key, out TValue value);
+        new bool TryGetValue(TKey key, out TValue value);
+        TValue IReadOnlyDictionary<TKey, TValue>.this[TKey key] => this[key];
+        IEnumerable<TKey> IReadOnlyDictionary<TKey, TValue>.Keys => Keys;
+        IEnumerable<TValue> IReadOnlyDictionary<TKey, TValue>.Values => Values;
+        bool IReadOnlyDictionary<TKey, TValue>.ContainsKey(TKey key) => ContainsKey(key);
+        bool IReadOnlyDictionary<TKey, TValue>.TryGetValue(TKey key, out TValue value) => TryGetValue(key, out value);
     }
-    public interface ISet<T> : ICollection<T> {
+    public interface ISet<T> : ICollection<T>, IReadOnlySet<T> {
-        bool IsProperSubsetOf(IEnumerable<T> other);
+        new bool IsProperSubsetOf(IEnumerable<T> other);
-        bool IsProperSupersetOf(IEnumerable<T> other);
+        new bool IsProperSupersetOf(IEnumerable<T> other);
-        bool IsSubsetOf(IEnumerable<T> other);
+        new bool IsSubsetOf(IEnumerable<T> other);
-        bool IsSupersetOf(IEnumerable<T> other);
+        new bool IsSupersetOf(IEnumerable<T> other);
-        bool Overlaps(IEnumerable<T> other);
+        new bool Overlaps(IEnumerable<T> other);
-        bool SetEquals(IEnumerable<T> other);
+        new bool SetEquals(IEnumerable<T> other);
// Adding this new member is required so that there's a most specific Contains method on ISet<T> since ICollection<T> and IReadOnlySet<T> define it too
+        new bool Contains(T value) => ((ICollection<T>)this).Contains(value); 
+        bool IReadOnlySet<T>.Contains(T value) => ((ICollection<T>)this).Contains(value);
+        bool IReadOnlySet<T>.IsProperSubsetOf(IEnumerable<T> other) => IsProperSubsetOf(other);
+        bool IReadOnlySet<T>.IsProperSupersetOf(IEnumerable<T> other) => IsProperSupersetOf(other);
+        bool IReadOnlySet<T>.IsSubsetOf(IEnumerable<T> other) => IsSubsetOf(other);
+        bool IReadOnlySet<T>.IsSupersetOf(IEnumerable<T> other) => IsSupersetOf(other);
+        bool IReadOnlySet<T>.Overlaps(IEnumerable<T> other) => Overlaps(other);
+        bool IReadOnlySet<T>.SetEquals(IEnumerable<T> other) => SetEquals(other);
+    }
 }
```

**Comment by MichalStrehovsky** [comment]
> Higher up in the thread somebody posted how to patch the loader to get it to auto-resolve those diamonds with a hinting attribute. So you can in fact reduce the breakage points to zero.

That's not true: https://github.com/terrajobst/ReadOnlyCollectionInterfaces/pull/1 doesn't involve conflict in default implementations. There's no diamond and even no default method in user code. The proposed throwaway loader feature would not fix that.

(I'm calling the feature throwaway because it can only be used if there's no other code using the new attribute marking things as fallback default implementations: we can only use it for one release - once this "fallback implementation" feature ships, we'll need "fallback of a fallback" attribute to use it again because now one can have fallbacks in user code. The feature is the type loader equivalent of [topmost window problem](https://devblogs.microsoft.com/oldnewthing/20110310-00/?p=11253))

**Comment by TylerBrinkley** [comment]
I'm not saying I'm advocating for this hinting attribute but couldn't you specify a priority property on the attribute to solve that issue and then reserve the highest priority for framework usage only?

**Comment by terrajobst** [comment]
I don't think the two failure modes that @MichalStrehovsky pointed out change my rationale why I think we should consider this change:

1. **Binding to existing virtual members**. This is illustrated by https://github.com/terrajobst/ReadOnlyCollectionInterfaces/pull/1.
2. **Ambiguous implementation when combined with user-defined DIMs**. This is illustrated by https://github.com/terrajobst/ReadOnlyCollectionInterfaces/pull/2.

The second one feels very rare because DIMs in user-defined types in the larger library ecosystem feel rare. I don't have data on this yet, but I plan to crawl nuget.org and report back how many there are. It's not gonna be zero, but it's also not gonna rise to 10-30% either. I suspect the use is less than 1%.

The first one also feels rare: How often will people have signatures from members in the `IReadOnlyXxx` interfaces that are `virtual` but unrelated to the semantics of the collection type? It is of course possible to construct them, but it doesn't seem very likely.
 
This is the totality of all members defined by `IReadOnlyXxx` interfaces that would have that problem:
 
```C#
int Count { get; }
T this[int index] { get; }
TValue this[TKey key] { get; }
IEnumerable<TKey> Keys { get; }
IEnumerable<TValue> Values { get; }
 
bool ContainsKey(TKey key);
bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value);
```

I would think if a type extends any of the mutable collection interfaces and defines these members as virtual members than their semantics would likely match those of the underlying `IReadOnlyXxx` interfaces.

Now, of course I can't conclusively proof any of this. But I think my reasoning is plausible enough to give this experiment a try early in .NET 9 and see what (if anything) breaks as a result of this.

**Comment by bartonjs** [comment]
> How often will people have signatures from members in the IReadOnlyXxx interfaces that are virtual but unrelated to the semantics of the collection type?

Presumably that's analyzable.  At least in the linked case, since it's a) it has explicitly implemented ICollection.Count, and b) has a (different) public member named Count.  In that case, IReadOnlyCollection.Count will be considered implicitly implemented, and not lifted from the DIM on ICollection.

**Comment by terrajobst** [comment]
@bartonjs 

> Presumably that's analyzable.

Maybe but I'll struggle with writing down what I would be looking for. There are many ways one could share the logic/semantics without necessarily binding to the same method across all v-table slots.

**Comment by joshudson** [comment]
Incidentally it's also fixable. Define a new attribute for default interface mehods that means "do not look for public methods to replace this with, but only methods declared to implement the interface.

I can tell that random public methods of the same name as interface methods don't necessarily impute themselves onto newly added interface methods because the opposite behavior exists in VB.NET; adding a (non-default) interface method results in a load error in VB.NET until the method is implemented in the class implementing the interface.

**Comment by MichalStrehovsky** [comment]
> Maybe but I'll struggle with writing down what I would be looking for. There are many ways one could share the logic/semantics without necessarily binding to the same method across all v-table slots.

The low tech approach could be just about looking at the list of methods. This would have to be explicit interface method implementation and the C# compiler mangles them in a distinct way.

If there's a method with name `System.Collections.Generic.ICollection\<.*\>.get_Count` in the base hierarchy, and another method with name `get_Count` (without the prefix) that returns an int, the class is likely susceptible to this problem. One could restrict to just virtual method (if `get_Count` is not virtual, it would not be a binary breaking change, but it would still be a source breaking change because the C# compiler is going to "helpfully" mark the method virtual if it sees it could implement an interface).

> Incidentally it's also fixable. Define a new attribute for default interface mehods that means "do not look for public methods to replace this with, but only methods declared to implement the interface.

If I understand it right, the fix would cause `class Foo : ICollection<T> { public int Count { ... } }` to not have `IReadOnlyCollection.Count` implemented by `Foo`, but by the default method on `ICollection` even in the future when we already shipped `ICollection : IReadOnlyCollection`. That sounds like a worse problem than the original problem. The default implementations are sub-optimal. They cause double interface call in spots that could just be one interface call. They also create an interface dispatch hotspot in the default implementation. Dynamic PGO is going to have a difficult time getting rid of them (because the callsite will likely be superpolymorphic). We want to run the convenience default implementations as little as possible.

**Comment by eiriktsarpalis** [comment]
If we want to run this experiment this year it would probably need to be merged before the end of January in time for Preview 1. Any volunteers willing to source an implementation?

**Comment by joshudson** [comment]
If nobody else steps up by Jan 2, I'll try it; but an earlier attempt would be better. Sorry, I cannot do it this month.

**Comment by terrajobst** [comment]
> Sorry, I cannot do it this month.

No need to apologize for not being able to volunteer your time to an OSS project, especially over the holiday period :-)

**Comment by TylerBrinkley** [comment]
I'll try my hand at it with #95830.

**Comment by eiriktsarpalis** [comment]
Change got reverted in https://github.com/dotnet/runtime/pull/101644, reopening to keep track of next steps.

**Comment by mikernet** [comment]
What are the next steps?

**Comment by RenderMichael** [comment]
With respect to C++/CLI (I’ve never used it but it sounds neat), this change is not a breaking change in any meaningful way, and is only breaking on C++/CLI due to a limitation in their type system.

 I’m not sure if it’s considered a bug to be fixed or a new feature that requires some cycles to design and implement, but I do hope this can be resolved on C++/CLI’s side and the change go through again. It would be a colossal disappointment if this change gets denied.

**Comment by huoyaoyuan** [comment]
Previously DIM in static interface method were blocked by C++/CLI too and temporarily reverted until RTM.
Hopefully C++/CLI can unblock on this too. I'm OK to see this change delayed to .NET 10, but unwilling to see it blocked forever.

**Comment by joshudson** [comment]
On considering the failure mode in C++/CLR; I am reasonably convinced it is not an ABI breaking change but only an API breaking change (that is, it only fails at compile time). That being said, the required source fix is _ugly_ and something should be done.

On further analysis, this seems to be a bug in the C++/CLR compiler. The problem happened upon adding `IList : IReadOnlyList` and a type was found that implements `IList` but doesn't expose a public method called `getCount` itself.

If the type follows CLR rules, it doesn't have a public method called `getCount`; therefore the type must follow C++ rules. On synthesizing the C++ view of the combined type I got:

```C++
#include <iostream>

class IEnumerable {
};

class IReadOnlyList: public virtual IEnumerable {
public:
	int getCount() { return 3; }
};

class IList: public virtual IEnumerable {
public:
	int getCount() { return 3; }
};

class MyList: public virtual IList {};

int main()
{
	MyList l;
	std::cout << l.getCount() << "\n";
}
```

Which builds and runs; the C++/CLR compiler is behaving as though the change imputed the following change to the code: `class MyList: public virtual IList, public virtual IReadOnlyList {};` where `MyList` only declares an implementation of `IList`

**Comment by KennethHoff** [comment]
Is there intention to try to get this back in time for a .Net 9 Preview? Is it being delayed to .Net 10+, or is it dead? 

The various discussions I've seen/heard seem to indicate that this is just a temporary issue when C++/CLI, but how temporary are we talking? I hope it's not as temporary as the lack of adoption for C++ modules 😅

**Comment by terrajobst** [comment]
It's not dead yet but it depends on whether or not we can control the failure cases. At this point .NET 9 is extremely unlikely.

**Comment by eiriktsarpalis** [comment]
Given we only have a month left of .NET 9 development, it is too late for a change carrying so much risk. Moving to 10.0.0

**Comment by KennethHoff** [comment]
I hope that, come November already (maybe early next year to allow time for Net9 to cook) they'll try to merge it in again

**Comment by jhudsoncedaron** [comment]
If you read the thread upwards; they have to come up with a solution to the problem that the C++/CLR compiler has a hard time of it. It looks like a deficiency in that compiler, but it's not going to be a trivial fix.

**Comment by terrajobst** [comment]
> Why not just do this already? 😔

First, https://www.youtube.com/watch?v=OP4CKn86qGY

It's on the docket for .NET 10. I agree that it's a huge quality of life improvement for C#, but we also need to make sure not to break everything else. And that requires coordination.

**Comment by KennethHoff** [comment]
This is heavily violating Liskov Substitution principle

**Comment by thomaslevesque** [comment]
It doesn't make sense. You don't want to expose modification methods on the read-only interface, even if they're not implemented.

**Comment by huoyaoyuan** [comment]
> is there a definitive list of protected APIs that cannot be modified due to C++/CLI constraints?

No. MSVC can change to fit new features, for example the static interface method used in .NET 8 requires msvc 14.38+(VS17.8+).

Everything used by the core BCL types can affect C++/CLI.

**Comment by RenderMichael** [comment]
Is there a tracking issue for the MSVC improvements which would enable this feature? I would love to give it a thumbs-up and I'm sure others would as well.

**Comment by rickbrew** [comment]
One of the big reasons I ported all of my C++/CLI interop glue code over to C# was because of all the problems I kept running into with new runtime and C# features causing compile errors (for C++/CLI code), or preventing me from using things like `Span<T>` and `unmanaged`.

**Comment by YoshiRulz** [comment]
> It doesn't make sense. You don't want to expose modification methods on the read-only interface, even if they're not implemented.

It would be a little at odds with the names of the interfaces, but they've always had weird semantics, and in practice that's how immutable collections are implemented: `Array` (and `ImmutableList<T>` if upcast) implement `IList`/`IList<T>` by throwing on mutation, `FrozenDictionary<K, V>` implements `IDictionary<K, V>` by throwing on mutation, and so on.

(If I may digress, Kotlin made the right choice by defining `interface MutableCollection<E> : Collection<E>`. Something else it has is [`@UnsafeVariance`](https://github.com/JetBrains/kotlin/blob/v2.0.0/core/builtins/native/kotlin/Collections.kt#L54) for allowing e.g. `contains(E)` in an interface that's `<out E>`, which is a feature I think .NET should copy.)

**Comment by thomaslevesque** [comment]
@YoshiRulz 
The whole point of introducing read-only interfaces was to only expose a subset of the methods of the writable interfaces, without the members causing mutation (and also benefit from covariance). If the read-only interface inherits from the writable interface, it's no longer a subset. In fact, since it doesn't introduce any new members, it might as well not exist at all, because it's effectively the same as the writable interface. Also, it can no longer be covariant (because T appears in input position in the method signatures).

> If I may digress, Kotlin made the right choice by defining `interface MutableCollection<E> : Collection<E>`.

I'm not sure I understand this argument. This is the same as `IList<T>` inheriting from `IReadOnlyList<T>`, but you suggest doing the opposite.

**Comment by RenderMichael** [comment]
“read only” has always been a misnomer, it’s always really mean “readable”. The ship has long since sailed on naming them `IReadableList<T>` etc, so these are the names we have to live with.

**Comment by jhudsoncedaron** [comment]
@YoshiRulz : If you want interfaces for things that cannot change; we have IImmutable* interfaces now.

**Comment by DrkWzrd** [comment]
It is setted in stone what will be the final API shape? What will it be?

**Comment by KennethHoff** [comment]
Sorry for pinging, but @terrajobst in [this comment](https://github.com/dotnet/runtime/issues/31001#issuecomment-2420500978) and especially in the .Net Api Review (IIRC) insinuated that in order for this to have any chance of merging it has to be done very early in development, which is where we're at now (Some could argue even this is too late; There's only like half a year or so until feature freeze after all).

Is there any news here? Would be nice to have some closure here (Be it good or bad).

I continue to use `IReadOnly*` interfaces but I semi-regularly encounter scenarios where some API uses `IList<T>` or `ICollection<T>` and it just breaks (requiring me to either use `ToArray`/`ToList`, or change all APIs up to the instantation point to use `IList<T>`/`ICollection<T>`) - let alone the potential performance problems of using `IReadOnly*` in conjunction with LINQ and similar, so maybe it's time to just let go of the idealized `IReadOnly*` and embrace accidental mutation.

**Comment by TylerBrinkley** [comment]
Yeah, I'm kind of curious who's responsible for this coordination Immo mentioned in readying it for .NET 10. Is it @eiriktsarpalis or @tarekgh as listed in the project it's attached to? If no-one's assigned to it within Microsoft we can't expect this to get done.

**Comment by tarekgh** [comment]
CC @jeffhandley

**Comment by Sergio0694** [comment]
We're starting to plan for some changes into [CsWinRT](https://github.com/microsoft/CsWinRT), and would love to know whether we could rely on these new interface implementations to be in .NET 10, as that would affect how we handle projecting/marshalling collection interface types to and from native code. Are there any updates on this from the C++/CLI side? Thank you!

**Comment by Sergio0694** [comment]
Mh, not sure why the auto-close didn't work -- this was just completed by #115802 🚀

**Comment by weitzhandler** [comment]
Related #23337.

**Comment by GerardSmit** [comment]
The PR has been reverted in #116497. I think this issue should be opened again?

The implementation will be blocked until https://github.com/fsharp/fslang-suggestions/issues/1430 has been fixed 🥲

**Comment by huoyaoyuan** [comment]
I thought we would have aligned with F# folks far before this. We really need to collaborate closer.

**Comment by tannergooding** [comment]
As has been mentioned in the thread numerous times, this is a very complex change due to it touching a core interface type that's been around for 20 years. A lot of the potential breaks are only found very late due to the subtleties around them.

This wasn't due to a lack of collaboration, but rather due to the mentioned complexities.

F# works fine in every other tested scenario around DIMs, there was a particular edge case missed that was triggered for this case instead. Even if F# adjusts itself to not break in this particular scenario, it is still a high risk change with a chance that it cannot be done due to breaking major 1st or 3rd party codebases which may be doing things via reflection or other tooling and which will break due to the change.

It is entirely possible this is a change that cannot ever be made due to the negative impact it would have on the ecosystem and that is something we knew was possible going into this. -- CC. @jkotas

**Comment by hez2010** [comment]
IMO we should try again at an early phase of .NET 11 instead of giving up at all, probably in .NET 11 preview 1, so that it can spread to enough places to verify the impacts to 3rd codebases.

**Comment by tannergooding** [comment]
As indicated above, it is very likely this is an impossible change to make due to how it impacts the ecosystem. Nothing can be done until the F# issue is resolved, at a minimum.

Beyond that it would still need to get done again to produce a private build. That private build would need to be shared out with major 1st and 3rd party codebases for validation/feedback, and it is very likely that one or more of them would also be broken at which point the change would be shut down for good.

We "can" do that work, but it's unlikely to pay off and it is dependent on other teams prioritizing and completing their own relevant work, so there's nothing further for the .NET Libraries team to drive until that occurs.

**Comment by rickbrew** [comment]
Well I for one certainly appreciate the persistence, I'm really looking forward to this hopefully landing eventually

**Comment by joshudson** [comment]
> As indicated above, it is very likely this is an impossible change to make due to how it impacts the ecosystem.

I'm not quite sure how you're coming to the conclusion that it is likely.

Theorem: this particular diamond has no logically unresolvable constructs, only constructs the loader doesn't know how to resolve.

Any offender must fall into one of these forms:

1) A compiler (including runtime IL generators)
2) A C++/CLR component
3) A runtime dependency on ICollection not inheriting from IReadOnlyCollection

All remaining cases are patchable even if the source code for the offending codebase is lost. I am reasonably certain all remaining cases are mechanically patchable, but it defeats my ability to write an autopatcher.

------

I've been wondering if the Linq cases should be inverted now; that is change all incidents of ICollection, IList, etc. to the IReadOnly equivalents. The remaining classes that implement ICollection but not IReadOnlyCollection should be low usage; to the point where that breaking change is managable in the community at large (it's just a performance loss not a code-doesn't-run case).

This would also enable a fallback plan: ICollection2: ICollection, IReadOnlyCollection, IList2: ICollection2, IReadOnlyList, etc; and then walk through ASP.NET codebase and find every method that takes ICollection, IList, etc. and add ICollection2 or IReadOnlyCollection overloads. Yes, anybody can write down the interface declarations but it doesn't *do any good* unless they're added to the BCL types including array itself. (This is the actual chokepoint; better interfaces can't be provided because of special code in Array and Linq explicitly recognizing HashSet.)

------

I've experimented with applying changes in a private build and dogfooding the private build; however I've come to the conclusion that private builds aren't meant to be used.

1) dotnet run --project usesprivatebuild.csproj doesn't use the private build
2) dotnet build --project usesprivatebuild.csproj doesn't pick up the private reference assemblies even though the resulting build uses the private build
3) Visual Studio doesn't recognize it needs to use the private assemblies in intellisense.

I can't fix these problems. I can only produce an autopatcher that patches the installed copy of the runtime.

This change is worthy of running a private build of .NET; but the toolchain is too dependent on people not doing that.

**Comment by tannergooding** [comment]
> I'm not quite sure how you're coming to the conclusion that it is likely.

Experience in the ecosystem and the impact such changes have to it, as well as discussion with other members of the .NET Libraries and Runtime teams.

There can be trivial real diamond problems caused if someone has defined their own interface that provides a DIM for the `IROCollection` members and some type implements both that and `ICollection`. However, we believe this to essentially be a non-existent issue.

The actual problem is all of the tooling. This includes core tooling like C++/CLI, F#, or other language compilers; all of which *must* at least handle DIMs for this change to function. While the risk decreases for 3rd party languages, it is still a significant impact and potential upgrade blocker for things which underpin much larger codebases. It also includes other tooling, essentially anything that deals with IL and inspecting the metadata. This can trivially break any tooling which doesn't understand DIMs or resolving conflicts between such newslot members. It forces that handling to exist due to how core `ICollection` is and the likelyhood it is encountered in existing normal code.

There will probably be one more attempt, if the F# blocker can get resolved. But there is still a decent likelihood that the change cannot be done.

> I've been wondering if the Linq cases should be inverted now

We'd done previous analysis on this and the answer was "no". There are still far too many collections across the entire ecosystem that only implement ICollection and not the IReadOnlyCollection counterparts. It would be a net negative to the ecosystem.

> This would also enable a fallback plan: ICollection2

The API review team had already decided this is a non-starter. The benefit it brings to the ecosystem is not enough to justify the overall worse UX. Developers can simply implement both interfaces on their readonly collection types and ensure that `ICollection.IsReadOnly` returns `true`. An analyzer that helps flag users that implement only the `rocollection` interfaces is a far better solution in such a world.

> I've experimented with applying changes in a private build and dogfooding the private build; however I've come to the conclusion that private builds aren't meant to be used.

.NET has a set of environment variables that need to be set for private builds. It's the same environment variables that are used by the `dotnet-install` scripts and anyone using nightly builds in a CI environment or similar: https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#net-sdk-and-cli-environment-variables

We would ensure appropriate instructions are given for using any private build we provide to the relevant parties, as we have done many times throughout the lifetime of .NET. It is something the team has experience around.

**Comment by eiriktsarpalis** [comment]
I gather from @T-Gro that the F# issue [has now been addressed](https://github.com/fsharp/fslang-suggestions/issues/1430). We might consider reviving the PR.

**Comment by weltkante** [comment]
Note the [comment in the PR](https://github.com/dotnet/runtime/pull/125004#issuecomment-3981492211) (quoting here for visibility):

> This change cannot be taken through dotnet/runtime. It must, due to risk, be taken through dotnet/dotnet to ensure the entire end to end product builds and works as expected.
> 
> However, there's also some internal discussions going on now as to whether this is something we can take as another potential break has come up and there are still concerns on the overall risk.
> 
> While there is no exact ETA on when we will have an update I do expect it will be "soon". If we determine that one more attempt is worthwhile, then the changes to dotnet/dotnet will be opened by a member of the .NET team; otherwise, a member of the .NET team will close the issue as inactionable with the relevant reasoning.
